nixos/modules/services/networking/trust-dns.nix

```diff
@@ -0,0 +1,175 @@
+  );
+
+  zoneType = lib.types.submodule ({ config, ... }: {
+    options = with lib; {
+      zone = mkOption {
+        type = types.str;
+        description = mdDoc ''
+          Zone name, like "example.com", "localhost", or "0.0.127.in-addr.arpa".
+        '';
+      };
+      zone_type = mkOption {
+        type = types.enum [ "Primary" "Secondary" "Hint" "Forward" ];
+        default = "Primary";
+        description = mdDoc ''
+          One of:
+          - "Primary" (the master, authority for the zone).
+          - "Secondary" (the slave, replicated from the primary).
+          - "Hint" (a cached zone with recursive resolver abilities).
+          - "Forward" (a cached zone where all requests are forwarded to another resolver).
+        '';
+      };
+      file = mkOption {
+        type = types.either types.path types.str;
+        default = "${config.zone}.zone";
+        defaultText = literalExpression ''"''${config.zone}.zone"'';
+        description = mdDoc ''
+          Path to the .zone file.
+          If not fully-qualified, this path will be interpreted relative to the `directory` option.
+          If omitted, defaults to the value of the `zone` option suffixed with ".zone".
+        '';
+      };
+    };
+  });
+in
+{
+  options = {
+    services.trust-dns = with lib; {
+      enable = mkEnableOption (lib.mdDoc "trust-dns");
+      package = mkOption {
+        type = types.package;
+        default = pkgs.trust-dns;
+        defaultText = "pkgs.trust-dns";
+        description = mdDoc ''
+          Trust-dns package to use.
+          Only `bin/named` need be provided: the other trust-dns utilities (client and resolver) are not needed.
+        '';
+      };
+      quiet = mkOption {
+        type = types.bool;
+        default = false;
+        description = mdDoc ''
+          Log ERROR level messages only.
+          This option is mutually exclusive with the `debug` option.
+          If neither `quiet` nor `debug` are enabled, logging defaults to the INFO level.
+        '';
+      };
+      debug = mkOption {
+        type = types.bool;
+        default = false;
+        description = mdDoc ''
+          Log DEBUG, INFO, WARN and ERROR messages.
+          This option is mutually exclusive with the `debug` option.
+          If neither `quiet` nor `debug` are enabled, logging defaults to the INFO level.
+        '';
+      };
+      settings = mkOption {
+        description = lib.mdDoc ''
+          Settings for trust-dns. The options enumerated here are not exhaustive.
+          Refer to upstream documentation for all available options:
+          - [Example settings](https://github.com/bluejekyll/trust-dns/blob/main/tests/test-data/test_configs/example.toml)
+        '';
+        type = types.submodule {
+          freeformType = toml.type;
+          options = {
+            listen_addrs_ipv4 = mkOption {
+              type = types.listOf types.str;
+              default = [];
```

Comment:
probably not, actually. listening on localhost makes sense for services that run behind a reverse proxy like nginx, but DNS servers aren't usually proxied like that. "127.0.0.1" wouldn't let trust-dns service queries from external machines, which is what the user probably wants. actually, it's common for machines to run their own DNS "stub resolvers" on 127.0.0.1 port 53 (systemd-resolved does this), and trust-dns would get in the way there.

i should document that trust-dns accepts `"*"` to listen on all addresses. if this works alongside systemd-resolved, it might be an OK default too.

i think i'll throw in an `openFirewall` option also.

Reply:
i've made trust-dns listen on all interfaces by default, as that's the same behavior we use for [bind](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/bind.nix#L176).

on the other hand, i decided not to add `openFirewall`, as it'd be difficult to write the rule which reliably opens all ports the user wants, and only those ports: some deployments will be just UDP port 53, others TCP and UDP, still more will add DNS over TLS (default 853), etc.

i don't have a strong preference on either of these though, happy to change if there's disagreement.

Reply:
This makes sense to me.